File: ios/LowerBackStretching/Audio/Tracks.swift
import Foundation

/// Catalogue of audio assets shipped with the app. Each entry maps a
/// persisted preference value to a bundle resource name and a
/// user-visible label.
///
/// `resourceName` is looked up via `Bundle.main.url(forResource:withExtension:)`
/// at play time — if the MP3 isn't in the bundle, the AudioController
/// silently falls back to no-op. See `Resources/Audio/README.md` for
/// file requirements.

enum MusicTrack: String, CaseIterable {
    case none, calm, lofi, piano

    var storageValue: String { rawValue }

    static func fromStorage(_ value: String?) -> MusicTrack {
        MusicTrack(rawValue: value ?? "") ?? .none
    }

    var resourceName: String? {
        switch self {
        case .none:  return nil
        case .calm:  return "music_calm"
        case .lofi:  return "music_lofi"
        case .piano: return "music_piano"
        }
    }

    var displayName: String {
        switch self {
        case .none:  return "None"
        case .calm:  return "Calm"
        case .lofi:  return "Lofi"
        case .piano: return "Piano"
        }
    }
}

enum AmbientTrack: String, CaseIterable {
    case none, rain, forest, ocean

    var storageValue: String { rawValue }

    static func fromStorage(_ value: String?) -> AmbientTrack {
        AmbientTrack(rawValue: value ?? "") ?? .none
    }

    var resourceName: String? {
        switch self {
        case .none:   return nil
        case .rain:   return "ambient_rain"
        case .forest: return "ambient_forest"
        case .ocean:  return "ambient_ocean"
        }
    }

    var displayName: String {
        switch self {
        case .none:   return "None"
        case .rain:   return "Rain"
        case .forest: return "Forest"
        case .ocean:  return "Ocean"
        }
    }
}

enum ChimeTrack: String, CaseIterable {
    case none, bell, ding, drop

    var storageValue: String { rawValue }

    static func fromStorage(_ value: String?) -> ChimeTrack {
        ChimeTrack(rawValue: value ?? "") ?? .none
    }

    var resourceName: String? {
        switch self {
        case .none: return nil
        case .bell: return "chime_bell"
        case .ding: return "chime_ding"
        case .drop: return "chime_drop"
        }
    }

    var displayName: String {
        switch self {
        case .none: return "None"
        case .bell: return "Bell"
        case .ding: return "Ding"
        case .drop: return "Drop"
        }
    }
}

enum AudioDefaults {
    static let musicVolume: Float = 0.4
    static let ambientVolume: Float = 0.6
}

File: android/app/src/main/java/com/lowerbackstretching/audio/Tracks.kt
package com.lowerbackstretching.audio

/**
 * Catalogue of audio assets shipped with the app. Each entry maps a
 * persisted preference value to a `res/raw/` resource name and a
 * user-visible label.
 *
 * `resName` is looked up via `resources.getIdentifier(...)` at play
 * time — if the file is missing from `res/raw/`, the AudioController
 * silently falls back to no-op (the build still succeeds; the feature
 * just plays nothing). See `android/app/AUDIO_FILES.md` for file requirements.
 */

enum class MusicTrack(
    val storageValue: String,
    val resName: String?,
    val displayName: String,
) {
    NONE("none", null, "None"),
    CALM("calm", "music_calm", "Calm"),
    LOFI("lofi", "music_lofi", "Lofi"),
    PIANO("piano", "music_piano", "Piano");

    companion object {
        fun fromStorage(value: String?): MusicTrack =
            entries.firstOrNull { it.storageValue == value } ?: NONE
    }
}

enum class AmbientTrack(
    val storageValue: String,
    val resName: String?,
    val displayName: String,
) {
    NONE("none", null, "None"),
    RAIN("rain", "ambient_rain", "Rain"),
    FOREST("forest", "ambient_forest", "Forest"),
    OCEAN("ocean", "ambient_ocean", "Ocean");

    companion object {
        fun fromStorage(value: String?): AmbientTrack =
            entries.firstOrNull { it.storageValue == value } ?: NONE
    }
}

enum class ChimeTrack(
    val storageValue: String,
    val resName: String?,
    val displayName: String,
) {
    NONE("none", null, "None"),
    BELL("bell", "chime_bell", "Bell"),
    DING("ding", "chime_ding", "Ding"),
    DROP("drop", "chime_drop", "Drop");

    companion object {
        fun fromStorage(value: String?): ChimeTrack =
            entries.firstOrNull { it.storageValue == value } ?: NONE
    }
}

object AudioDefaults {
    const val MUSIC_VOLUME = 0.4f
    const val AMBIENT_VOLUME = 0.6f
}
